h2o-algos/src/main/java/hex/glm/GLMModel.java

```diff
@@ -721,10 +802,29 @@ public final double likelihood(double yr, double ym) {
     public GLMWeights computeWeights(double y, double eta, double off, double w, GLMWeights x) {
       double etaOff = eta + off;
       x.mu = linkInv(etaOff);
+//      x.mu = x.mu==0?hex.glm.GLMTask.EPS:x.mu;
       double var = variance(x.mu);//Math.max(1e-5, variance(x.mu)); // avoid numerical problems with 0 variance
       double d = linkDeriv(x.mu);
-      x.w = w / (var * d * d);
-      x.z = eta + (y - x.mu) * d;
+      if (_family.equals(Family.negativebinomial)) {
+        double invSum = 1.0/(1+_theta*x.mu);
+        double d2 = linkInvDeriv(x.mu);
+        if (y>0 && (x.mu>0)) {
+          double sumr = 1.0+_theta*y;
+          d = (y/(x.mu*x.mu)-_theta*sumr*invSum*invSum) * d2 * d2 + (sumr*invSum-y/x.mu) * linkInvDeriv2(x.mu); //CHECKED-log/CHECKED-identity
+          x.w = w*d;
+          x.z = eta + (y-x.mu) *invSum * d2/(d*x.mu); // CHECKED-identity
+        } else if (y==0 && x.mu > 0) {
+          d = linkInvDeriv2(x.mu)*invSum-_theta*invSum*invSum*d2*d2; // CHECKED
+          x.w = w*d;
+          x.z = eta - invSum*d2/d;
+        } else {
+          x.w = 0;
+          x.z = 0;
+        }
+      } else {
+        x.w = w / (var * d * d);  // formula did not quite work with negative binomial
+        x.z = eta + (y - x.mu) * d;
```

Comment:
All the likelihood/deviance, hessian and gradient calculations can be found in Negative Binomial Regression paper from NCSS.com chapter 326 pages 2 to 3.